src/auth0-sso-login.js

```diff
@@ -6,13 +6,55 @@ export class windowInteraction {
   static updateWindow(url) {
     window.location = url;
   }
+
+  static setTimeout(func, delay) {
+    return window.setTimeout(func, delay);
+  }
+
+  static clearTimeout(timeoutId) {
+    window.clearTimeout(timeoutId);
+  }
 }
 
+// static localStorage interaction, mostly wrapped for enabling easier mocking through unit tests
+export class localStorageInteraction {
```

Comment:
I suggest pulling this out into its own class, we could `import` it here and then `mock` it, rather than `stubbing`.  It would provide us assurances the methods don't change.  It also prevents accidental usage of the `localStorage` in the unit tests if we forget to stub.

Since we never use a `key` other than the `tokenExpiresAtKey` I'm almost thinking that should be injected into that class and hide the ability for the library to use different keys.

Something like:
```javascript
const tokenExpiresAtKey = 'cimpress.auth0.tokenExpiresAt';
export default class localStorageIntegration {
  getTokenExpiry() {
    let tokenExpiresAtString = localStorage.getItem(tokenExpiresAtKey);
     return tokenExpiresAtString ? JSON.parse(tokenExpiresAtString) : null;
  },
  setTokenExpiry(value) {
    localStorage.setItem(tokenExpiresAtKey, JSON.stringify(value));
  }
  removeTokenExpiry() {
    localStorage.removeItem(tokenExpiresAtKey);
  }
}
```

Reply:
New question, why even use the `localStorage` here.  Is there a scenario where we would want to remember the last refresh after the site was closed?  Doesn't auth have to happen again anyway?  If that is the case, in memory is a better solution.  You get the free `removeItem` and don't have to worry about the store.

Reply:
I agree with the first suggestion.

There is a scenario. We have to consider that the library is not triggering initial login on its own. It is triggered from the client (using `ensureLoggedIn`). When the client logic checks the validity of the token outside of this library and therefore client skips calling `ensureLoggedIn` on page load, we need a way to schedule the renewal. Step by step:
1. Application gets token using `ensureLoggedIn`
2. Library schedules renewal in `tokenRefreshed`
3. User does page refresh (library looses scheduled renewal)
4. Library schedules renewal in `constructor`
5. Application will not call `ensureLoggedIn` as it has valid token from the previous login (no call to `tokenRefreshed`). Therefore, the 4th step was required.

Reply:
If we can't guarantee that `ensureLoggedIn` is called, then we can't guarantee that `constructor` is called.  Why not always require `ensureLoggedIn` is called to get all the functionality.  Should we investigate the case where `ensureLoggedIn` isn't not supposed be called.  If we could somehow still assume it would be (possible parameter, or other check) then that would reduce the complexity.

Reply:
Actually it is not `ensureLoggedIn` directly, it's about `tokenRefreshed` which is scheduling the renewal after getting new token.

Two questions:
1. Is it feasible to get a new token on every application load? I don't think so.
2. How can we guarantee that token renewal will be registered after application load? I thought `constructor` is the best place we can have. Maybe more aggressively, we can schedule token renewal on `window.event('load')`

Reply:
I think we can easily adjust the renewal scheduling to allow calling `ensureLoggedIn`. What really prevents us from doing so is that it will try to get a new token.

The library doesn’t have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

Reply:
My reason for this inquiry: constructors with side effects tell me something is off.
To answer your questions:

> 1. Is it feasible to get a new token on every application load? I don't think so.

What code will be executed in this process?  I.e. does application load mean, new tab, F5, hard refresh etc...?  In that case, I would say we actually need to call `ensureLoggedIn` because even if the token is valid, the SSO session may not be.  The token should almost be considered a session specific object, and not saved in a clients application.  It may provide some improvements, such as not showing the `login` button temporarily, but the source of truth is the Auth0 SSO session.

> 2. How can we guarantee that token renewal will be registered after application load? I thought constructor is the best place we can have.

I wouldn't worry about this.  We can always say, to use this library execute, is there a reason we need to be as aggressive?  Is there a reason we need to avoid `ensureLoggedIn`
```javascript
import auth0Manager from 'auth0-sso-login';
auth0Manager.ensureLoggedIn({ avoidAuth0LockWidget: true });
```

Can we follow that line of questioning?  Why does the library need to:
> have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

Reply:
If we require `ensureLoggedIn` after importing this library, we increase complexity on the client side. What if their application load does not require auth0 token immediately, but only upon entering a specific protected route. If clients protect the route with `ensureLoggedIn`, they will get a new token on every route entry.

>have access to the token, therefore it cannot check if there is already one and prevent getting a new token.

Knowing the token would solve this. Just let clients call `ensureLoggedIn` whenever they need to know that a valid token is present. This library can decide, whether to get a new one or not.

Reply:
(Is one more line of code extra complexity? I think the trade off to not having unintended side-effects on an import statement or constructor is worth the trade off)

>Knowing the token would solve this. Just let clients call ensureLoggedIn whenever they need to know that a valid token is present. This library can decide, whether to get a new one or not.

I agree with this, so figuring out how to get that in is important, but I'm not sure that answers the question of why we need to use the `localStorage`.

I see what the problem is now.  I would expect that `ensureLoggedIn` to be called every time the application is opened, but not always force the user to the auth0Lock.  in other words, we should always try to get a token if possible, but if login isn't required then don't show the lock.

For example take [this google site](https://support.google.com/adwords/?hl=en#topic=3119071) for instance, while login isn't required, you can see from the screen shot that I'm still recognized.  I was expecting this library to perform the same action.

![image](https://user-images.githubusercontent.com/5056218/34034171-a5e8665e-e17d-11e7-83c4-b4094423e4dc.png)


I.e. try to get the token always, and then let the application decide how to handle missing a token.  In other words there are there things that need to be done:
* Get a token if available => Should always happen
* Schedule the refresh => Should always happen
* Get the Auth0 Popup => Should only happen when navigating to a page that requires a valid token.

With the new implementation we are moving the first two into the third one.  But I posit that these are actually separate actions.

If we call `ensureLoggedIn` to be synonymous with the `Get the Auth0 Popup if necessary` that explains why we want to move to the constructor the scheduling of the refresh.

Instead I would suggest:
* On application start call the following code:
```javascript
  import auth0Manager from 'auth0-sso-login';
  const auth0Manager = new Auth0Manager();
  auth0Manager.ensureLoggedIn();
```
* Then on each route change that requires a token
```javascript
   auth0Manager.ensureLoggedIn({ enableLockWidget: true });
```
We could also go the direction towards pulling the logic into the constructor as this PR is taking it, but we'll should still handle getting a new token at the same time as the refresh if possible without triggering the lock widget.

Reply:
I see it shaping :) What you say with the parameter to `ensureLoggedIn` makes sense. We can get rid of the `localStorage` as I agree that new application load should get a new token as well. So storing the expiry date in memory is sufficient.

There is still one deal breaker though. Calling ` auth0Manager.ensureLoggedIn({ enableLockWidget: true })` will always get a new token. How to prevent this? It again goes to the direction if this library should store the token on its own or not.